### PR TITLE
Coupled flow transport

### PR DESCRIPTION
**Pull request by lipnikov**
See last messages in the discussion below for summary of changes.

Original:
"Daniil and I implemented two new PKs for coupled (matrix-fracture) flow and transport and related infrastructure. 

Epetra maps were replaced with Epetra block maps which allows us now to use heterogeneous distribution of DOFs, e.g. one or two fluxes on different mesh faces. Supermaps were generalized to assemble block systems with different sizes of blocks. New operators were implemented to support non-PDE coupling terms.

Regions of fractures extracted from an Exodus mesh using labels are handled correctly when cells are requested on the derived mesh.  

Native documentation was extended, userguide got one new benchmark test."

### CODE REVIEW

**Comment by lipnikov** [comment]
Daniil will comment on "multi_domain" parameter, since he was working on this design.

**Comment by lipnikov** [comment]
It is 2, we have two vectors with different maps then laying them out as [u_0, u_1, ..., u_N, v_0, v_1, ... v_M]. We do not want to assume that v is a subset of u.

**Comment by ecoon** [comment]
Ok, then what do you need BlockMap in CompositeVector for at all?  Isn't block map intended to interleave them, i.e. [u_0_0, u_1_0, u_1_1, u_1_2, u_2_0, u_2_1, ... ] (where u_i_j is element i, dof j)?

**Comment by lipnikov** [comment]
In matrix, we have one flux and one pressure on all faces except for fracture faces, where we need two fluxes and two pressures.

File: src/operators/Op.hh [hunk 1]
```diff
@@ -112,16 +118,19 @@ class Op {
   virtual void ApplyMatrixFreeOp(const Operator* assembler,
           const CompositeVector& X, CompositeVector& Y) const = 0;
 
+  virtual void ApplyMatrixFreeOpVariableDOFs(const Operator* assembler,
+          const CompositeVector& X, CompositeVector& Y) const {};
```
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/operators/Op.hh]: Either keep it pure or place an empty routine to all derived classes.

File: src/data_structures/SuperMap.cc [hunk 2]
```diff
@@ -154,19 +215,23 @@ SuperMap::CreateIndices_(const std::string& compname, int dofnum, bool ghosted)
...
       indices[i] = offset + dofnum + i*num_dof;
     }
-
+      
+    int MyPID = BaseGhostedMap(compname)->Comm().MyPID();
     int ghosted_offset = ghosted_offsets_.at(compname);
+
     for (int i=nentities_owned; i!=nentities; ++i) {
       indices[i] = ghosted_offset + dofnum + (i-nentities_owned)*num_dof;
+      //if (MyPID==1) std::cout << "ghost "<<i<<" "<<indices[i] <<"\n";
```
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/data_structures/SuperMap.cc]: Thanks. This usually goes away with time.

**Comment by ecoon** [comment]
That makes sense as to why you need both concepts then.

That leaves multi_domain.  It seems to me that a ton of interfaces were changed simply to pass along a single boolean whose only job is to tell SuperMap, "despite the fact that this is called the 'cell' component, it isn't the same map as the other vector's 'cell' component."

If that's true, can't we just write a SuperMap that checks Map->SameAs() or some other quick check on the maps themselves?  You have the maps, you know if they are the same or not.

File: src/operators/Operator_FaceCell.cc [hunk 3]
```diff
@@ -167,35 +168,134 @@ int Operator_FaceCell::ApplyMatrixFreeOp(const Op_SurfaceFace_SurfaceCell& op,
...
+        lid_c[k] = offset + (first + m) * num_dof_faces + col_pos;
+        k++;
+      }
+    }
+
+    int cell_dof_size = cell_gh_map->ElementSize(c);
+    int first = cell_gh_map->FirstPointInElement(c);
+    for (int m = 0; m != cell_dof_size; ++m) {
+      lid_r[k] = cell_row_offset + (first + m) * num_dof_cells + row_pos;
+      lid_c[k] = cell_row_offset + (first + m) * num_dof_cells + col_pos;
```
Comment (ecoon) [inline src/operators/Operator_FaceCell.cc]: No no, this and code like this is the problem.  This defeats the entire purpose of SuperMap.  The point of SuperMap is that it provides a list of the indices associated with a given component.  No where outside of SuperMap should the ordering of the Map be known.  Here you are assuming you understand the internals of how SuperMap lays out its data.  Please don't do that -- that is the reason for the GhostIndices() methods.

Instead, refactor SuperMap to provide the right indices for a given map.  This is the entire point of SuperMap, don't work around it, just generalize support for BlockMap into that capability.

This code should look like this:


```
/* ******************************************************************
* Visit methods for symbolic assemble: FaceCell
****************************************************************** */
void Operator_FaceCell::SymbolicAssembleMatrixOp(const Op_Cell_FaceCell& op,
                                                 const SuperMap& map, GraphFE& graph,
                                                 int my_block_row, int my_block_col) const
{
  std::vector<int> lid_r(2*cell_max_faces + 1);
  std::vector<int> lid_c(2*cell_max_faces + 1);
    
  // ELEMENT: cell, DOFS: cell and face
  auto face_row_inds = map.GhostIndices("face", my_block_row);
  auto face_col_inds = map.GhostIndices("face", my_block_col);
  auto cell_row_inds = map.GhostIndices("cell", my_block_row);
  auto cell_col_inds = map.GhostIndices("cell", my_block_col); 

  // NOTE: this is not a limitation, it is required in the current design too.
  // It probably does not need to actually be checked here, as SameAs() is
  // somewhat expensive.
  // ASSERT(map.BaseGhostedMap("face", my_block_row).SameAs(map.BaseGhostedMap("face", my_block_col));
  // ASSERT(map.BaseGhostedMap("cell", my_block_row).SameAs(map.BaseGhostedMap("cell", my_block_col));
  
  Teuchos::RCP<const Epetra_BlockMap> face_gh_map = map.BaseGhostedMap("face", my_block_row);
  Teuchos::RCP<const Epetra_BlockMap> cell_map = map.BaseMap("cell", my_block_row);


  AmanziMesh::Entity_ID_List faces;
  for (int c = 0; c != ncells_owned; ++c) {
    mesh_->cell_get_faces(c, &faces);
    int nfaces = faces.size();

    int k = 0;
    for (int n = 0; n != nfaces; ++n) {
      int f = faces[n];
      int face_dof_size = face_gh_map->ElementSize(f);
      int first = face_gh_map->FirstPointInElement(f);

      for (int m = 0; m != face_dof_size; ++m) {
        lid_r[k] = face_row_inds[first + m];
        lid_c[k] = face_col_inds[first + m];
        k++;
      }
    }

    int cell_dof_size = cell_gh_map->ElementSize(c);
    int first = cell_gh_map->FirstPointInElement(c);
    for (int m = 0; m != cell_dof_size; ++m) {
      lid_r[k] = cell_row_inds[first + m];
      lid_c[k] = cell_col_inds[first + m];
      k++;
    }
    
    ierr |= graph.InsertMyIndices(k, lid_r.data(), k, lid_c.data());
  }
  AMANZI_ASSERT(!ierr);
}

```

**Comment by dasvyat** [comment]
I'm little lost in the flow of the discussion, so I'd like to start from the very beginning. At first, we'd like to make CompositeVectorSpaces independent on mesh and mesh entities. Make CompositeVectorSpaces rely on BlockMaps. Maybe it is not completely independent at the moment because I did not replace old interfaces with new ones. These map are provide by "user" who wants to create this CVS, either PK or Operator. Since  CompositeVectorSpaces doesn't rely on entities anymore cell-0 and cell-1 may have different maps. The multi_domain parameter is design to distinguish two situations: (False) components with the same name has the same map; (True) components with the same name has different maps. For the fracture-matrix problems "cell" for fracture and "cell" for matrix are two different components. For the pressure-temperature example multi_domain=True correspond to [p1, p2 .... p_N, T1,T2, ... T_N], while multi_domain=False correspond to [p1, T1, p2, T2, .....p_N, T_N].

**Comment by ecoon** [comment]
Right.  There was a design flaw in the original SuperMap whereby anything that was called "cell" was assumed to be the same map, no matter what mesh/whatever it came from.  This design flaw was called out in the documentation (you probably saw it).

This is problematic for you especially, and why you need to put these two "cell" maps into different blocks.

Instead of passing a bool, let me push some changes to this branch which deal with this case.  We HAVE the maps, we can check if they are the same or not.  I'll do this tonight after my kids go to bed, it should be a quick change.

This will allow SuperMap to detect when two vectors both have "cell" but they have different maps.  There is nothing that says that "cell" vector 1 MUST be interleaved with "cell" vector 2.  We can instead check that, when "cell" vector 2 is added, if it is the same as 1, it is interleaved, but if not, it gets its own block.

Then, when you ask for `GhostIndices("cell", 2)`, you should be able to get the right thing, even though it is not interleaved with 1.  Then you should be able to write code like the above example that I sketched out.  This should remove all need for "multi_domain."

File: src/operators/Operator_Cell.cc [hunk 4]
```diff
@@ -127,13 +136,26 @@ void Operator_Cell::SymbolicAssembleMatrixOp(const Op_Cell_Cell& op,
...
+
+  std::string cell_name_row = "cell";
+  std::string cell_name_col = "cell";
+  int row_pos = my_block_row;
+  int col_pos = my_block_col;
+
+  if (multi_domain) {
+    cell_name_row = cell_name_row + "-" + std::to_string(my_block_row);
+    cell_name_col = cell_name_col + "-" + std::to_string(my_block_col);
+    row_pos = 0;
```
[the remark this answers is not in the document]
Comment (dasvyat) [reply src/operators/Operator_Cell.cc]: But you can't always assume that you can block components. Right now you either block all or nothing. Theoretically you can avoid blocking by using appropriate different components names.

**Comment by lipnikov** [comment]
Hope your algorithms works with the following case: 2 "cell" copies for vector 1 and 3 "cell" copies for vector 2.

**Comment by dasvyat** [comment]
Ethan, I agree with you that it would much clean if we can move some stuff from Operator_(Face_Cell) to SuperMap. But approach that you are describing will not work, at least as is.  We assume that heterogeneous map may also have multiple dof. For example, on a bulk mesh we define a BlockMap such that there are 2 map elements on some internal faces corresponding to fractures. Moreover, (hypothetically, not right now) we like to use 3 dof for each map element for flux and 1 dof for each map element for pressure. So old notation of num_dofs_ is still valid. So GhostIndices should have the third argument. Do you agree?

**Comment by ecoon** [comment]
What do you mean by copies?  This is getting a bit complex because of the variety of options here.  We have options for three levels of variations:

- BlockMap with ElementSize > 1  
- CompositeVectorSpace with SetComponent(name, KIND, num_dofs > 1)
- TreeVectorSpace with multiple CompositeVectorSpaces, each with the same map of the same name.

To be clear, the current (master) version of SuperMap was intended to handle only the 3rd.  See:

https://github.com/amanzi/amanzi/blob/master/src/data_structures/SuperMap.cc#L272
https://github.com/amanzi/amanzi/blob/master/src/data_structures/SuperMap.cc#L278

Effectively here the concept was that two PKs (think flow + energy) would both need MFD's face + cells, and we wanted to order them as 

```[p_c0, T_c0, ..., p_cN-1, T_cN-1, p_f0, T_f0, ..., p_fM-1, T_fM-1, p_cN, T_cN, ..., p_cN+NGC-1, T_cN+NGC-1, p_fM, T_fM, ..., p_fM+NGF-1, T_fM+NGF-1]```

where c0 ... cN-1 refer to cells, where N = Number of OWNED cells, f is faces with M = number of OWNED faces, then NGC and NGF are number of ghosted cells and faces respectively.  Effectively this was reordering to put all the ghost cells at the end, interleaving the cell and faces.  But critically, in the use case that you constructed a SuperMap from a TreeVector of identically-mapped blocks, each with only 1 dof, this implied that the "block number" was equivalent to the "dof number."

**Comment by lipnikov** [comment]
By "copies" I meant the second case.

**Comment by ecoon** [comment]
> So old notation of num_dofs_ is still valid. So GhostIndices should have the third argument. Do you agree?

Yes.  See: https://github.com/ecoon/amanzi/blob/ecoon/supermap/src/data_structures/SuperMapWrapper.hh#L74

**Comment by ecoon** [comment]
Ok, I think the changes to SuperMap should cover what you need here.  As I understand it, the main thing that still needs to happen to operators to enable the physics changes here is to update the `Operator_*::ApplyMatrixFreeOp()` and `Operator_*::SymbolicAssembleMatrixOp()` and `Operator_*::AssembleMatrixOp()` methods to loop over Points instead of Elements.  

_Should this be done to every routine?_

It seems to me that, from a design standpoint, if you're going to support BlockMap access patterns, it should be done for all discretizations, not just for a few.  And it is straightforward to do it -- it just requires getting NumPoints() and FirstPointInElement() and expanding the list of indices to assemble into.  Both of these things are just inlined array accesses, and therefore should be quite cheap.  This updates all the assemblers to look like the one above (which I edited to match the implemented SuperMap interface as Daniil pointed out would be necessary).

Note that if we do find it is expensive to check NumPoints(), etc, then we can optimize based on whether `Epetra_BlockMap::ElementSize() == 1`.

**Comment by lipnikov** [comment]
Using a relatively complex code for simple Operator_* is overkill. What are trying to achieve? There are many way to generalized Operators, e.g. limiting them to a  region.  Should we always "upgrade" the simple code? We have two options: (1) new Operators, (2) new visit patterns. I am leaning towards option (2).

**Comment by ecoon** [comment]
We have the following things that affect how a schema's indexing LID (e.g. element cell 6) gets mapped into the LIDs its local matrix gets assembled into:

- block structure **SuperMap**
- discretization (i.e. MFD needs cell + all neighboring faces) **Op / Schema**
- Map vs BlockMap (use NumPoints() and FirstPointInElement() ** ?? ** 
- extracted/parent mesh (cell 6 is actually on a surface mesh, and it needs to be assembled into the spot of the corresponding parent's face) **Op**, e.g. **Op_SurfaceFace_SurfaceCell**
- potentially subsets/regions or other modifiers

I guess sticking it in the Op is fine?  It seems that this ought to be decomposable a bit better.  Putting it in the Op then means that we need (if we intend to support) twice as many visits or an if statement in every visit.

**Comment by lipnikov** [comment]
A factory of assemblers i.e. operators?

File: src/data_structures/CompositeVectorSpace.hh [hunk 5]
```diff
@@ -115,6 +132,19 @@ public:
...
 
+  CompositeVectorSpace*
+  SetComponent(std::string& name,
+               Teuchos::RCP<const Epetra_BlockMap> mastermap,
+               Teuchos::RCP<const Epetra_BlockMap> ghostmap,
+               int num_dof);
+
+  CompositeVectorSpace*
+  SetComponents(const std::vector<std::string>& names,
+                const std::vector<AmanziMesh::Entity_kind> locations,
```
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/data_structures/CompositeVectorSpace.hh]: Fixed recently on master branch.

File: src/data_structures/CompositeVectorSpace.hh [hunk 6]
```diff
@@ -115,6 +132,19 @@ public:
                 const std::vector<AmanziMesh::Entity_kind>& locations,
                 const std::vector<int>& num_dofs);
 
+  CompositeVectorSpace*
+  SetComponent(std::string& name,
```
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/data_structures/CompositeVectorSpace.hh]: Fixed on master branch

File: src/data_structures/CompositeVectorSpace.cc [hunk 7]
```diff
@@ -19,7 +30,7 @@ namespace Amanzi {
 
 // constructor
 CompositeVectorSpace::CompositeVectorSpace() :
-  owned_(false), mesh_(Teuchos::null), ghosted_(false) {}
+  owned_(false), mesh_(Teuchos::null), ghosted_(false) {};
 
 // copy constructor
 CompositeVectorSpace::CompositeVectorSpace(const CompositeVectorSpace& other) :
```
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/data_structures/CompositeVectorSpace.cc]: Modified as suggested.

File: src/data_structures/SuperMap.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/data_structures/SuperMap.cc]: My suggestion is to move it to a separate ticket, see discussion below.

File: src/data_structures/CompositeVectorSpace.cc [hunk 8]
```diff
@@ -123,11 +139,50 @@ CompositeVectorSpace::AddComponent(std::string name,
...
+
+CompositeVectorSpace*  
+CompositeVectorSpace::AddComponent(std::string name,
+                                   Teuchos::RCP<const Epetra_BlockMap> mastermap,
+                                   Teuchos::RCP<const Epetra_BlockMap> ghostmap,
+                                   int num_dof) {
+
+  std::vector<std::string> names(1,name);
+  std::vector<int> num_dofs(1,num_dof);
+  std::vector<AmanziMesh::Entity_kind> locations(1, AmanziMesh::UNDEFINED);
```
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/data_structures/CompositeVectorSpace.cc]: Resolved on the master branch.

File: src/operators/Op.hh (hunk 1 above)
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/operators/Op.hh]: My suggestion is to move this issue to a separate ticket, see discussion below.

File: src/operators/Operator_FaceCell.cc (hunk 3 above)
Comment (lipnikov) [reply src/operators/Operator_FaceCell.cc]: Used this example to modify the code.

File: src/operators/Operator.hh [hunk 9]
```diff
@@ -298,90 +312,98 @@ class Operator {
   // visit methods for symbolic assemble
   virtual void SymbolicAssembleMatrixOp(const Op_Cell_FaceCell& op,
           const SuperMap& map, GraphFE& graph,
-          int my_block_row, int my_block_col) const;
+          int my_block_row, int my_block_col, bool multi_domain) const;
```
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/operators/Operator.hh]: Resolved on the master branch.

File: src/operators/OperatorUtils.hh [hunk 10]
```diff
@@ -45,15 +45,17 @@ int CopySuperVectorToCompositeVector(const SuperMap& map, const Epetra_Vector& s
 
 // Nonmember TreeVector to/from Super-vector
 // -- simple schema version
-int CopyTreeVectorToSuperVector(const SuperMap& map, const TreeVector& cv, Epetra_Vector& sv);
-int CopySuperVectorToTreeVector(const SuperMap& map, const Epetra_Vector& sv, TreeVector& cv);
-int AddSuperVectorToTreeVector(const SuperMap& map, const Epetra_Vector& sv, TreeVector& cv);
+int CopyTreeVectorToSuperVector(const SuperMap& map, const TreeVector& cv, bool multi_domain, Epetra_Vector& sv);
+int CopySuperVectorToTreeVector(const SuperMap& map, const Epetra_Vector& sv, bool multi_domain, TreeVector& cv);
+int AddSuperVectorToTreeVector(const SuperMap& map, const Epetra_Vector& sv, bool multi_domain, TreeVector& cv);
```
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/operators/OperatorUtils.hh]: Resolved on the master branch.

File: src/operators/Operator_Cell.cc (hunk 4 above)
[the remark this answers is not in the document]
Comment (lipnikov) [reply src/operators/Operator_Cell.cc]: Resolved on the master branch.

**Comment by lipnikov** [comment]
Summary: everything was fixed except for mapping schema's LIDs to matrix LIDs. This requires major development and better to be pushed to a separate ticket. We need first to polish last two commits, plus a new awaiting PR on stronger coupling of discretization (whetstone) with operator's schema.

**Comment by ecoon** [comment]
That's effectively what we have now -- a way of constructing a collection of objects (Op) that specifically names all of these things, then a single implementation location to stick each Op in the given Operator.  One object per combination.

What I'm imagining (but realistically don't have the time to do) is a few helper classes which, when composed, unique describe how to assemble.  So, it might look something like:

```Op<CELL, FACE & CELL>;```
is the standard MFD op with no special bells or whistles.

```OpHelper_Points<Op<CELL, FACE & CELL> >```
 is the  standard MFD op on a BlockMap that needs points, not elements.

So note that you would only implement:

```template<Base_Op> OpHelper_Points;```
which would work with also `Op<FACE,CELL>`, etc, and take any basic Op that provided a schema and do so with points instead of elements.

Similarly, you might have:

```template<Base_Op> OpHelper_Extracted;```

which would take a Base_Op, but lift its schema up a dimension (from surface mesh to mesh) and get the entity_parent()'s LID instead.

Then composing, i.e.

```OpHelper_Extracted< OpHelper_Points< Op<CELL, FACE & CELL> > >;```

should be possible.

Then an Operator would provide a templated visitor pattern:

```template<typename Op_type> Operator::SymbolicAssembleOp( Op_type& op );```
which would get the SuperMap indices and magically make it all work.

This may end up more complex than is useful, especially if we only use a few combinations of these.  And unless there is a clear view of how these are composed, we'd end up implementing just as many "special case" combinations.  So it would take more work than I think I want to spend right now to even convince myself it is doable in a useful way.

So after all that, I'd say just make another Op...

File: src/operators/OperatorUtils.cc [hunk 11]
```diff
@@ -334,5 +336,78 @@ CreateBoundaryMaps(Teuchos::RCP<const AmanziMesh::Mesh> mesh,
 }
 
 
+/* ******************************************************************
+* TBW
+****************************************************************** */
+std::pair<Teuchos::RCP<const Epetra_BlockMap>, Teuchos::RCP<const Epetra_BlockMap> >
```
Comment (ecoon) [inline src/operators/OperatorUtils.cc]: This method is already implemented in src/data_structures: https://github.com/amanzi/amanzi/blob/master/src/data_structures/CompositeVectorSpace.hh#L36

Can you use that one instead?

File: src/state/State.cc [hunk 12]
```diff
@@ -932,10 +932,15 @@ void State::Initialize() {
...
 
 void State::Initialize(Teuchos::RCP<State> S) {
+  auto vo = Teuchos::rcp(new VerboseObject("State", state_plist_)); 
+  Teuchos::OSTab tab = vo->getOSTab();
+  
   for (FieldMap::iterator f_it = fields_.begin();
        f_it != fields_.end(); ++f_it) {
     Teuchos::RCP<Field> field = f_it->second;
     Teuchos::RCP<Field> copy = S->GetField_(field->fieldname());
+    *vo->os() << "processing field \"" << f_it->first << "\"\n";
```
Comment (ecoon) [inline src/state/State.cc]: Please stick a verbosity on this, i.e. 

```if (vo->os_OK(Teuchos::VERB_???))```

to make it clear how to supress this.  I would prefer it was at least `VERB_HIGH` and maybe even `VERB_EXTREME` as this seems like debug messages to me.

Comment (lipnikov) [reply src/state/State.cc]: More like statistics. But it is also useful for smart debugging.  It was added to have at least an rough idea where the code silently dies.

File: src/operators/OperatorUtils.cc (hunk 11 above)
Comment (lipnikov) [reply src/operators/OperatorUtils.cc]: Good catch. Eliminated operators's version of getMaps.

File: src/state/State.cc (hunk 12 above)
Comment (lipnikov) [reply src/state/State.cc]: selected VERB_HIGH

**Pull request merged**